### ISSUE TITLE
filtered search on multiple fields, should be easy when you know how do it well ;)

### ISSUE THREAD

**Issue by alotela** [karmi/retire#826]
Hi,

I don't know if this is the best way to do it, and I don't think so, so I'm asking for help if someone as a better way of doing this:

mapping in model :
indexes :label, type: 'string'
indexes :creator, type: 'string'
indexes :shop do
   indexes :label, type: 'string'
end

query in controller:
query do
 filtered do
    query do string "label:_#{search_term}_ OR creator:_#{search_term}_ OR shop.label:_#{search_term}_", default_operator: 'AND' end
    filter :and, selected_filters #=> no matter here
  end
end

I need to search on these 3 fields. I also need the and operator for words, since "red cat" must return all products/shops with "red cat" in fields, and not "red" or "cat" in fields.

I know \* queries are slow and I should use indexes. But for now, I'm still experimenting, so if someone can help me doing it faster (or a better way), it's welcome. I thought about boolean filtered queries, but no luck for now (tried multi_match, but it does not seem to be the right way)

thank you.

**Comment by karmi** [comment]
Well, you should use filters only for _filtering_ data, ie. answering questions "is this tagged with _foo_?". Queries not only match data, but also compute relevancy.

So, in your case, I suppose you correctly set up filters for stuff like tags, authorship, status, etc.

You can rewrite the "foo OR bar" query quite easily into a `boolean` query, see integration test for inspiration, https://github.com/karmi/tire/blob/master/test/integration/boolean_queries_test.rb

**Comment by alotela** [comment]
ok,

I tried it before, it was one of my options:

MyModel.tire.search do
  query do
    boolean do
      should do string "label:_#{search_term}_" end
      should do string "shop.label:_#{search_term}_" end
      .....
    end
    filtered do
      filter :and, selected_filters
    end
  end
  facets...(I need data to be filtered for all facets)
  sort
end

but I think I have an error into my query ? ;) Nothing returned

**Comment by alotela** [comment]
I finished with this :

``` ruby
query do
 filtered do
    query do 
      boolean do
                  should { string "label:*#{search_term}*", default_operator: 'AND' }
                  should { string "creator:*#{search_term}*", default_operator: 'AND' }
                  should { string "shop.label:*#{search_term}*", default_operator: 'AND' }
                end
              end
              filter :and, selected_filters
            end
          end
```

which seems to be good. I needed to scope boolean block into query block.
Is this the good solution ?

**Comment by karmi** [comment]
Yes, note that you can use the `match` query instead of `string` query in the shoulds, or, use the `match` query in a "multi_match" mode -- see integration tests for examples.

**Comment by alotela** [comment]
ok, I tried match, but don't know how to use wildcard. ES doc says it's not possible...?

I think I found the way I wanted to do it:
query do
            filtered do
              query do 
                string "_#{search_term}_", default_operator: 'AND', fields: [:label, :creator, :'shop.uri_label']
              end
              filter :and, selected_filters
            end
          end

what do you think ?

Next, I need to work on removing widlcard to have faster searches with analyzed fields.
